### ISSUE TITLE
Boost-Context based builds are broken

### ISSUE THREAD

**Issue by RichieSams** [RichieSams/FiberTaskingLib#8]
Functional tests all seg-fault. The unit tests have managed to fix a number of other erros. However, this issue still remains.

Issue #13 shows that the seg-faults go away if we execute the tests in a specific order. This could mean that the shutdown code is missing something

**Comment by RichieSams** [comment]
The functional tests all segfault on the build server

**Comment by RichieSams** [comment]
Fixed the TLS errors with ce35163888294d3a5f39f23f17050123df4a0698
So all current units pass. However, the Functional Tests still cause a Seg Fault

**Comment by RichieSams** [comment]
With OSX builds now being tested, now we know it's just Linux builds that are failing. I still have no idea why.

**Comment by RichieSams** [comment]
Both clang and gcc builds fail on Linux

**Comment by RichieSams** [comment]
Try replacing moodycamel lockless queues with simple mutex'ed queues. 

Internally, moodycamel queues use thread local storage. This may be what is causing the undefined behavior in Linux builds

**Comment by JodiTheTigger** [comment]
Ok. Looks like we have some phantom resource running out.

I took out the producer_consumer google test and made it a stand alone program.
I limited it to two threads.

This will run fine on my machine:
```cpp
const uint kNumProducerTasks = 10u;
const uint kNumConsumerTasks = 20u;
```
But once I start increasing the number of tasks, then segfaults start creeping in (I run the test 100 times, and run that 10 times, see when it crashes).

210 tasks: ok
1010 tasks: ok
1510 tasks: ok
1535 tasks: ok
1536 tasks: crashes 3 / 10 times
1760 tasks: crashes 4 / 10 times
2010 tasks: crashes 7 / 10 times

**Comment by JodiTheTigger** [comment]
The 1535 test also crashes now and then after many runs, so it's not the threshold I was hoping it to be.

**Comment by JodiTheTigger** [comment]
Another thought is that it's actually some sort of thread race, and it's just more probable with more tasks running.

**Comment by JodiTheTigger** [comment]
Oooh, I think I have a cause. I ran strace on my test program:

```
gettid()                                = 2662
gettid()                                = 2662
gettid()                                = 2662
sched_yield()                           = 0
sched_yield( <ptrace(SYSCALL):No such process>
+++ killed by SIGSEGV (core dumped) +++
fish: “strace ./FiberTaskingLib-test” terminated by signal SIGSEGV (Address boundary error)
```

**Comment by JodiTheTigger** [comment]
Anyway. It seems to be crashing inside `Fiber::~Fiber()` on the line:
```cpp
			m_stack->deallocate(m_stackContext);
```
```cpp
    void deallocate( stack_context & sctx) {
        BOOST_ASSERT( sctx.sp);
        BOOST_ASSERT( traits_type::minimum_size() <= sctx.size);
        BOOST_ASSERT( traits_type::is_unbounded() || ( traits_type::maximum_size() >= sctx.size) );

#if defined(BOOST_USE_VALGRIND)
        VALGRIND_STACK_DEREGISTER( sctx.valgrind_stack_id);
#endif

        void * vp = static_cast< char * >( sctx.sp) - sctx.size;
        std::free( vp);   // <----------- CRASHES HERE
    }
```

**Comment by JodiTheTigger** [comment]
LOL, I have no idea what I'm doing now, but this seems to solve the issue:

from 
```cpp
		if (m_stack) {
			m_stack->deallocate(m_stackContext);
```
to

```cpp
        if (m_stack) {

            static bool first_free_ignore = true;
            if (!first_free_ignore)
            {
                m_stack->deallocate(m_stackContext);
            }
            else
            {
                first_free_ignore = true;
            }
```

**Comment by JodiTheTigger** [comment]
I suspect that somehow it is trying to free the context of the main thread (lol, wut?)

**Comment by RichieSams** [comment]
Interesting..... 

The main thread fiber should have a nullptr stack, so the branch should take care of it. The other thing I can think of is something is calling the destructor more than once. The hack way to fix it is to null out the stack ptr after the first delete. Any subsequent deletes would pass through. But, that's not fixing the underlying problem

**Comment by JodiTheTigger** [comment]
Ok, so I wrote code to track context allocs and deallocs. Nope, no double frees there

It seems the free on the first stack is somehow invalid. It would imply the first context free is a double free. Tried to run with memory sanitisers to validate the hunch, but then I can't trigger the crash. Stupid heisenbug.

**Comment by JodiTheTigger** [comment]
(note: build with -DBOOST_USE_VALGRIND if valgrinding).

I'm 100% stumped. Valgrind reports  an invalid write of size 8 at `std::free( vp);` above. I have put breakpoints there, and that's the _first_ time free is called via deallocate. Who and where is free been called at that address before I get there? Or why does valgrind think it's invalid? I suspect Valgrind is getting confused with boost::contexdt somehow because it doesn't actually segfault.

The main thread (called from main) does have a null stack, so `FTLConvertFiberToThread(FTLGetCurrentFiber())` does nothing (when called in TaskScheduler::Quit() before waiting for the threads to finish).

There is no point nulling out the stack pointer after the first delete, because it crashes on the first delete.

I give up. I just don't know wtf is going on. :sniff:

**Comment by RichieSams** [comment]
Thank you again for trying. But yea, that's pretty much how I feel right now. I have absolutely no idea why its failing. I join you in your sorrow. ha ha /cry

**Comment by RichieSams** [comment]
Below is a snapshot of the output I get when I run on my Ubuntu VM. It will run infinitely until I manually kill the process.

It looks like there's some kind of corruption going on where we somehow manage to
move the execution pointer back to the beginning of the test execution main().

I have no idea what the "16607:    transferring control: " is. I tried googling it, but nothing yet. All I know is that 16607 is the unique id of the thread.


    [==========] Running 7 tests from 3 test cases.
    [----------] Global test environment set-up.
    [----------] 2 tests from FiberAbstraction
    [ RUN      ] FiberAbstraction.NestedFiberSwitch
    [       OK ] FiberAbstraction.NestedFiberSwitch (0 ms)
    [ RUN      ] FiberAbstraction.SingleFiberSwitch
    [       OK ] FiberAbstraction.SingleFiberSwitch (0 ms)
    [----------] 2 tests from FiberAbstraction (0 ms total)

    [----------] 3 tests from FunctionalTests
    [ RUN      ] FunctionalTests.Maze10x10
    [       OK ] FunctionalTests.Maze10x10 (7 ms)
    [ RUN      ] FunctionalTests.ProducerConsumer
    [       OK ] FunctionalTests.ProducerConsumer (1035 ms)
    [ RUN      ] FunctionalTests.CalcTriangleNum
    [       OK ] FunctionalTests.CalcTriangleNum (108 ms)
    [----------] 3 tests from FunctionalTests (1150 ms total)

    [----------] 2 tests from TLSAbstraction
    [ RUN      ] TLSAbstraction.GetSet
    [       OK ] TLSAbstraction.GetSet (5 ms)
    [ RUN      ] TLSAbstraction.MultipleThreadGetSet
    [       OK ] TLSAbstraction.MultipleThreadGetSet (76 ms)
    [----------] 2 tests from TLSAbstraction (81 ms total)

    [----------] Global test environment tear-down
    [==========] 7 tests from 3 test cases ran. (1231 ms total)
    [  PASSED  ] 7 tests.
    [==========] Running 7 tests from 3 test cases.
    [----------] Global test environment set-up.
    [----------] 2 tests from FiberAbstraction
    [ RUN      ] FiberAbstraction.NestedFiberSwitch
         16607:    
         16607:    transferring control: /home/adrian/.clion11/system/cmake/generated/451f95cd/451f95cd/Debug0/tests/FiberTaskingLib-test
         16607:    
    [       OK ] FiberAbstraction.NestedFiberSwitch (0 ms)
    [ RUN      ] FiberAbstraction.SingleFiberSwitch
    [       OK ] FiberAbstraction.SingleFiberSwitch (0 ms)
    [----------] 2 tests from FiberAbstraction (0 ms total)

    [----------] 3 tests from FunctionalTests
    [ RUN      ] FunctionalTests.Maze10x10
    [       OK ] FunctionalTests.Maze10x10 (12 ms)
    [ RUN      ] FunctionalTests.ProducerConsumer
    [       OK ] FunctionalTests.ProducerConsumer (1037 ms)
    [ RUN      ] FunctionalTests.CalcTriangleNum
    [       OK ] FunctionalTests.CalcTriangleNum (111 ms)
    [----------] 3 tests from FunctionalTests (1160 ms total)

    [----------] 2 tests from TLSAbstraction
    [ RUN      ] TLSAbstraction.GetSet
    [       OK ] TLSAbstraction.GetSet (8 ms)
    [ RUN      ] TLSAbstraction.MultipleThreadGetSet
    [       OK ] TLSAbstraction.MultipleThreadGetSet (112 ms)
    [----------] 2 tests from TLSAbstraction (120 ms total)

    [----------] Global test environment tear-down
    [==========] 7 tests from 3 test cases ran. (1280 ms total)
    [  PASSED  ] 7 tests.
    [==========] Running 7 tests from 3 test cases.
    [----------] Global test environment set-up.
    [----------] 2 tests from FiberAbstraction
    [ RUN      ] FiberAbstraction.NestedFiberSwitch
    [       OK ] FiberAbstraction.NestedFiberSwitch (0 ms)
    [ RUN      ] FiberAbstraction.SingleFiberSwitch
    [       OK ] FiberAbstraction.SingleFiberSwitch (0 ms)
    [----------] 2 tests from FiberAbstraction (0 ms total)

    [----------] 3 tests from FunctionalTests
    [ RUN      ] FunctionalTests.Maze10x10
         16607:    
         16607:    transferring control: /home/adrian/.clion11/system/cmake/generated/451f95cd/451f95cd/Debug0/tests/FiberTaskingLib-test
         16607:    
    [       OK ] FunctionalTests.Maze10x10 (13 ms)
    [ RUN      ] FunctionalTests.ProducerConsumer
    [       OK ] FunctionalTests.ProducerConsumer (1039 ms)
    [ RUN      ] FunctionalTests.CalcTriangleNum
    [       OK ] FunctionalTests.CalcTriangleNum (108 ms)
    [----------] 3 tests from FunctionalTests (1160 ms total)

    [----------] 2 tests from TLSAbstraction
    [ RUN      ] TLSAbstraction.GetSet
    [       OK ] TLSAbstraction.GetSet (12 ms)
    [ RUN      ] TLSAbstraction.MultipleThreadGetSet
    [       OK ] TLSAbstraction.MultipleThreadGetSet (132 ms)
    [----------] 2 tests from TLSAbstraction (144 ms total)

    [----------] Global test environment tear-down
    [==========] 7 tests from 3 test cases ran. (1304 ms total)
    [  PASSED  ] 7 tests.
         16607:    
         16607:    transferring control: /home/adrian/.clion11/system/cmake/generated/451f95cd/451f95cd/Debug0/tests/FiberTaskingLib-test
         16607:    
    [==========] Running 7 tests from 3 test cases.
    [----------] Global test environment set-up.
    [----------] 2 tests from FiberAbstraction
    [ RUN      ] FiberAbstraction.NestedFiberSwitch
    [       OK ] FiberAbstraction.NestedFiberSwitch (0 ms)
    [ RUN      ] FiberAbstraction.SingleFiberSwitch
    [       OK ] FiberAbstraction.SingleFiberSwitch (0 ms)
    [----------] 2 tests from FiberAbstraction (0 ms total)

**Comment by JodiTheTigger** [comment]
Just thought of a longshot: Try using _older_ versions of boost context, to see if the issue goes away.

(inspiration from http://boost.2283326.n4.nabble.com/coroutine-x86-msvc11-stack-corruption-td4657990.html)

**Comment by RichieSams** [comment]
Worth a shot!

**Comment by RichieSams** [comment]
I'll try that when I get home in 15 min

**Comment by RichieSams** [comment]
Boost builds fail on windows as well. But all tests pass on both Linux and Windows if we hard limit the scheduler to only use 1 thread. So this makes me think that the threads are somehow accessing the same fiber.

In addition, Windows builds with Win32 fibers, but TLS abstraction still work. So I don't believe it's due to the fake TLS. Perhaps it's the mix of two though???

**Comment by JodiTheTigger** [comment]
I found the bug for linux at least (I ended up writing my own fibre library, and ended up with the same bunch of issues. I used the depreciated ucontext lib, and once I've cleaned it all up I'll try again with boost context)

When a thread exits, it has to have the same context it had when it started.

**Comment by RichieSams** [comment]
Huh... that's a really interesting observation. Thanks! I coulf probably add some tracking of the first context, so we can switch back to the original before exiting.

**Comment by JodiTheTigger** [comment]
Yea, I spent 80% of my time trying to figure out how to switch back to the original context without adding warts all over my code. 

The way I did it was to save the original context out of the bounds of the normal algorithm, and then come shutdown time, run a special job that would figure out what thread it is, read the special context and switch straight to it, letting the thread exit (bypassing my alogrithm completly (lol, context switching is just an over-glorified goto statement)). 

The Exception was thread 0, which would just do nothing for the special job, then I would use the algorithms counter mechanism to quit out normally so that thread 0 would be the one that exited out.

My code is a debug mess atm, but hopefully in the next few weeks I can release it so you can see what I mean.

**Comment by RichieSams** [comment]
Fixed by commits 1837d45e152e9bf2742248f127434dfad7e118d8 to 013ee8cabd4ab39226e0aeb6c492b9a025ea2cc3
